src/geometry/WorkingGeometry.ts

```diff
@@ -149,45 +154,77 @@ export class WorkingGeometry {
-        const scalingVector: vec3 = this.getScalingVector(destinationVector, pullVector);
-        const scalingMatrix: mat4 = mat4.fromRotationTranslationScaleOrigin(
+    public freeformStretchTo(
+        pullPoint: vec3,
+        destinationPoint: vec3,
+        holdPoint: vec3 = vec3.create()
+    ) {
+        const moveHoldToOrigin = mat4.translate(
             mat4.create(),
-            quat.create(),
-            vec3.create(),
-            scalingVector,
-            holdPoint
+            mat4.create(),
+            vec3.negate(vec3.create(), holdPoint)
         );
-        this.transform(scalingMatrix);
+
+        const translatedPullPoint = vec3.sub(vec3.create(), pullPoint, holdPoint);
+        const translatedDestinationPoint = vec3.sub(vec3.create(), destinationPoint, holdPoint);
+
+        const rotPullToXAxis = mat4.fromQuat(
+            mat4.create(),
+            quat.rotationTo(
+                quat.create(),
+                vec3.normalize(vec3.create(), translatedPullPoint),
+                vec3.fromValues(1, 0, 0)
+            )
+        );
+
+        const scalePullToDestOnXAxis = mat4.scale(
+            mat4.create(),
+            mat4.create(),
+            vec3.fromValues(
+                vec3.len(translatedDestinationPoint) / vec3.len(translatedPullPoint),
+                1,
+                1
+            )
+        );
+
+        const rotXAxisToDestAxis = mat4.fromQuat(
+            mat4.create(),
+            quat.rotationTo(
+                quat.create(),
+                vec3.fromValues(1, 0, 0),
+                vec3.normalize(vec3.create(), translatedDestinationPoint)
+            )
+        );
+
+        const moveOriginToHold = mat4.translate(mat4.create(), mat4.create(), holdPoint);
+
+        // Assemble all the matrices!
+        const matrix = moveOriginToHold;
+        mat4.mul(matrix, matrix, rotXAxisToDestAxis);
+        mat4.mul(matrix, matrix, scalePullToDestOnXAxis);
+        mat4.mul(matrix, matrix, rotPullToXAxis);
+        mat4.mul(matrix, matrix, moveHoldToOrigin);
+
+        this.transform(matrix);
     }
 
     /**
-     * Scale by a given factor between 2 points.
+     * Scale by a given factor away from a given point.
      *
      * @param {number} factor: scaling factor.
-     * @param {vec3} pullPoint: point that will be scaled by factor away from holdPoint.
      * @param {vec3} holdPoint: point to scale from, default to true origin.
      */
-    public scaleByFactor(factor: number, pullPoint: vec3, holdPoint: vec3 = vec3.create()) {
-        const scalingDirection: vec3 = vec3.sub(vec3.create(), pullPoint, holdPoint);
-        const absScalingDirection: vec3 = vec3.fromValues(
-            Math.abs(scalingDirection[0]),
-            Math.abs(scalingDirection[1]),
-            Math.abs(scalingDirection[2])
-        );
+    public proportionalStretchTo(factor: number, holdPoint: vec3 = vec3.create()) {
```

Comment:
So I may have missed part of a conversation, but it looks like this function removed the pull point. The pull point allowed for scaling by a given factor on a given axis. It looks like proportionalStetchTo scales the object in every direction, which has its uses, but I can see the merits of keeping a function which can scale along a given axis by a certain factor.

Reply:
@abhimadan You mentioned that it "scales equally in all directions", does that mean the `holdPoint` of the object will no longer have the coordinates of `holdPoint` after scaling? It sounds like `holdPoint` becomes the "centre" of scaling whereas the expected behaviour is for it to be the single unmoving point.

Reply:
That's what `freefromStretchTo` is for :)

That function doesn't allow you to specify a scale by an explicit factor, but it does scale along an axis.

Reply:
I also got confused with a few things in my previous comments. The purpose of `scale` and `scaleByFactor` were to allow the user to do the same things but with different interfaces.
The purpose for each:
* With `scale` it's possible for the user to freeform alter the object however they like. If they want it to look a certain way they can specify a coordinate value and then scale their object. They can also skew their object using this function.
* With `scaleByFactor` it'll allow for them to make accurate changes to their model if they very much require units that are precise. The user should not be able to skew their function since the direction of the scale is always `pullPoint - holdPoint`, but otherwise it's the same as `scale`. At the time of the initial PR we decided to do skewing at a later time.

I think I got confused with the proportional scaling since our test only had that, but the initial purpose was not to make `scaleByFactor` be proportionally constrained.

Reply:
Right, but I think it would also be useful if the user could specify a stretch along a diagonal axis say, without needing to do the math themselves.

Reply:
Sorry about all the churn on this PR — I didn't follow the PR that brought in these changes, and the original implementation didn't seem correct to me because it could produce uninvertible matrices (which would mess up normal transformations).

So if we want non-uniform scaling here, the implementation will almost be the same as `freeformStretchTo`, except that step 3 (see the earlier comment explaining the implementation) will use the provided factor as the x-scale, and step 4 will rotate back to the same line as `pullPoint`. Is that OK with you @taimengliu and @pbardea?

Reply:
FWIW, I still think that having a uniform scaling option is useful, in case you don't want to shear your object.